File: backend/user_service/api_key_service.py
from pathlib import Path
import json
from typing import Optional, Dict
from database.encryption_service import EncryptionService

class APIKeyService:
    def __init__(self):
        self.encryption_service = EncryptionService()
        self.keys_file = Path(__file__).parent / 'encrypted_keys'
        self._initialize_storage()

    def _initialize_storage(self):
        """Initialize storage file if it doesn't exist"""
        if not self.keys_file.exists():
            self.keys_file.write_bytes(self.encryption_service.encrypt_data({}))

    def save_alpaca_keys(self, user_id: str, api_key: str, secret_key: str, is_paper: bool) -> bool:
        """Save encrypted Alpaca API keys for a user"""
        try:
            # Read existing data
            encrypted_data = self.keys_file.read_bytes()
            data = self.encryption_service.decrypt_data(encrypted_data)
            
            # Update with new keys
            data[user_id] = {
                'alpaca': {
                    'api_key': api_key,
                    'secret_key': secret_key,
                    'is_paper': is_paper
                }
            }
            
            # Encrypt and save
            encrypted_new_data = self.encryption_service.encrypt_data(data)
            self.keys_file.write_bytes(encrypted_new_data)
            return True
        except Exception as e:
            print(f"Error saving keys: {e}")
            return False

    def get_alpaca_keys(self, user_id: str) -> Optional[Dict]:
        """Retrieve Alpaca API keys for a user"""
        try:
            encrypted_data = self.keys_file.read_bytes()
            data = self.encryption_service.decrypt_data(encrypted_data)
            return data.get(user_id, {}).get('alpaca')
        except Exception as e:
            print(f"Error retrieving keys: {e}")
            return None

    def delete_alpaca_keys(self, user_id: str) -> bool:
        """Delete Alpaca API keys for a user"""
        try:
            encrypted_data = self.keys_file.read_bytes()
            data = self.encryption_service.decrypt_data(encrypted_data)
            
            if user_id in data and 'alpaca' in data[user_id]:
                del data[user_id]['alpaca']
                if not data[user_id]:  # If no other keys, remove user entirely
                    del data[user_id]
                
                encrypted_new_data = self.encryption_service.encrypt_data(data)
                self.keys_file.write_bytes(encrypted_new_data)
            return True
        except Exception as e:
            print(f"Error deleting keys: {e}")
            return False

File: backend-swift/Sources/App/Services/APIKeyService.swift
import Foundation

struct StoredKeyData: Codable {
    var apiKey: String
    var secretKey: String
}

actor APIKeyService {
    private let fileManager = FileManager.default
    private let keysURL: URL
    private let encryptionService: EncryptionService
    
    init(keysPath: String, encryptionService: EncryptionService) async throws {
        self.keysURL = URL(fileURLWithPath: keysPath)
        self.encryptionService = encryptionService
        
        // Initialize storage if needed
        try await initializeStorage()
    }
    
    private func initializeStorage() async throws {
        if !fileManager.fileExists(atPath: keysURL.path) {
            let emptyData = try JSONEncoder().encode([String: StoredKeyData]())
            let encryptedData = try encryptionService.encrypt(emptyData)
            try encryptedData.write(to: keysURL)
        }
    }
    
    func saveAlpacaKeys(userId: String, apiKey: String, secretKey: String) async throws {
        do {
            let encryptedData = try Data(contentsOf: keysURL)
            let decryptedData = try encryptionService.decrypt(encryptedData)
            
            // Parse and update data
            var storedData = try JSONDecoder().decode([String: StoredKeyData].self, from: decryptedData)
            
            storedData[userId] = StoredKeyData(apiKey: apiKey, secretKey: secretKey)
            
            // Save updated data
            let updatedData = try JSONEncoder().encode(storedData)
            let encryptedUpdatedData = try encryptionService.encrypt(updatedData)
            try encryptedUpdatedData.write(to: keysURL)
        } catch {
            print("Error saving keys: \(error)")
            throw error
        }
    }
    
    func getAlpacaKeys(userId: String) async throws -> AlpacaKeyPair? {
        do {
            let encryptedData = try Data(contentsOf: keysURL)
            let decryptedData = try encryptionService.decrypt(encryptedData)
            
            let storedData = try JSONDecoder().decode([String: StoredKeyData].self, from: decryptedData)
            
            guard let userData = storedData[userId] else {
                return nil
            }
            
            return AlpacaKeyPair(apiKey: userData.apiKey, secretKey: userData.secretKey)
        } catch {
            print("Error retrieving keys: \(error)")
            return nil
        }
    }
    
    func deleteAlpacaKeys(userId: String) async throws {
        do {
            let encryptedData = try Data(contentsOf: keysURL)
            let decryptedData = try encryptionService.decrypt(encryptedData)
            
            var storedData = try JSONDecoder().decode([String: StoredKeyData].self, from: decryptedData)
            storedData.removeValue(forKey: userId)
            
            let updatedData = try JSONEncoder().encode(storedData)
            let encryptedUpdatedData = try encryptionService.encrypt(updatedData)
            try encryptedUpdatedData.write(to: keysURL)
        } catch {
            print("Error deleting keys: \(error)")
            throw error
        }
    }
}

struct AlpacaKeyPair: Codable {
    let apiKey: String
    let secretKey: String
}
